server/src/test/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitorTests.java

```diff
@@ -227,4 +207,123 @@ protected void markIndicesReadOnly(Set<String> indicesToMarkReadOnly, ActionList
         monitor.onNewInfo(new ClusterInfo(allDisksOk, null, null, null));
         assertNull(listenerReference.get());
     }
+
+    public void testAutoReleaseIndices() {
```

Comment:
I think this test needs another case to assert that we do not auto-release an index if it has a shard on a node with no disk usage information (particularly if it _does_ have shards on nodes that have enough free disk usage). IOW if I comment out the following line from the implementation ...

```
markNodesMissingUsageIneligibleForRelease(routingNodes, usages, indicesNotToAutoRelease);
```

... then this test should fail. Today it passes.

I also think it should test some cases of auto-releasing where an index has multiple shards: we must not auto-release if one of the shards is on a node with enough free space if another shard of the same index is over the high watermark.